### PR TITLE
MINOR: Support streaming decompression of fetched records for new format

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
cc @ijuma @junrao @apurvam 

One of the nice benefits of the new message format is the ability to do streaming decompression of fetched data. In the old message format, we had to decompress each record batch in full before returning any data because we needed to find the last offset to determine the initial offset. But in the new format, we have everything we need in the batch header, so we can return records to the user as they are being decompressed. This works nicely with the consumer's `max.poll.records` config. We only decompress and deserialize as much data as we need to satisfy each `poll()` and we can leave the rest compressed. The benefit is reduced memory utilization (see also KIP-81).

**Review by apurvam — COMMENTED** [review]
Looks good to me overall. I spent a bit of time understanding the old code and the new, so I probably won't be able to point out corner cases. My only question is around testing: 

Do we have coverage testing compressed produce/fetch in the unit / integration / and system tests? 

If so, we probably want to run system tests against this branch. If not, we probably should add more tests to get coverage on these changes.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -844,29 +821,18 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
                 throw new IllegalStateException("Unexpected error code " + error.code() + " while fetching data");
             }
         } finally {
-            completedFetch.metricAggregator.record(tp, bytes, recordsCount);
+            if (error != Errors.NONE) {
+                completedFetch.metricAggregator.record(tp, bytes, recordsCount);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Could we just call `record(tp, 0, 0)` here and get rid of the `bytes` and `recordsCount` variables?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 2]
```diff
@@ -894,42 +860,102 @@ public void onAssignment(Set<TopicPartition> assignment) {
...
+                    }
+                    currentBatch = batches.next();
+                    maybeEnsureValid(currentBatch);
+                    records = currentBatch.iterator();
+                }
+
+                Record record = records.next();
+                maybeEnsureValid(record);
+
+                // skip any records out of range
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:931]: The old code is "record.offset() < position" while we are adding the `==` case as well. Was the old code a bug then?

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:931]: Also, we did not update the `nextFetchOffset` at each record as well. Why couldn't we still update just once as 

```
long nextOffset = partRecords.get(partRecords.size() - 1).offset() + 1;
```

outside?

**Review by guozhangwang — COMMENTED** [review]
LGTM overall, just a few clarification questions.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:931]: > The old code is "record.offset() < position" while we are adding the == case as well. Was the old code a bug then?

I think the checks are equivalent. In the old code, we did this to skip records:
```java
if (record.isControlRecord() || record.offset() < position) {
    skippedRecords = true;
}
```

Now we do this to _include_ records:
```java
// skip any records out of range
if (record.offset() >= nextFetchOffset) {
    nextFetchOffset = record.offset() + 1;

    // control records are not returned to the user
    if (!record.isControlRecord())
         return record;
}
```

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:931]: > Also, we did not update the nextFetchOffset at each record as well.

The main point is that now we have control records. When we encounter one, it is not returned to the user, but we still increment the consumer's offset to show that we have processed it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Ack

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 2 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:931]: Makes sense.

**Review by guozhangwang — APPROVED** [review]

**Comment by hachikuji** [comment]
Ran consumer system tests here: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-03-28--001.1490746004--hachikuji--streaming-compressed-iterator--fcdddf4/report.html.

**Pull request closed without merging**